
// .\Release\benchmarks.exe  --benchmark_repetitions=3 --benchmark_filter=Ephemeris                                                                     // NOLINT(whitespace/line_length)
// Benchmarking on 1 X 3310 MHz CPU
// 2015/06/14-18:27:09
// Benchmark                                              Time(ns)    CPU(ns) Iterations                                                                // NOLINT(whitespace/line_length)
// -------------------------------------------------------------------------------------                                                                // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMajorBodiesOnly               23768936371 23540550900          1                                 +1.00027592626789170e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMajorBodiesOnly               23454743838 23368949800          1                                 +1.00027592626789170e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMajorBodiesOnly               24137182211 24070954300          1                                 +1.00027592626789170e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMajorBodiesOnly_mean          23786954140 23660151667          1                                 +1.00027592626789170e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMajorBodiesOnly_stddev          278895456   298809680          0                                 +1.00027592626789170e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMinorAndMajorBodies           50229173397 50123121300          1                                 +1.00027592631378660e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMinorAndMajorBodies           48035647129 47892307000          1                                 +1.00027592631378660e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMinorAndMajorBodies           47750342106 47611505200          1                                 +1.00027592631378660e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMinorAndMajorBodies_mean      48671720878 48542311167          1                                 +1.00027592631378660e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemMinorAndMajorBodies_stddev     1107427500  1123664517          0                                 +1.00027592631378660e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemAllBodiesAndOblateness        59442244744 59202379500          1                                 +1.00027592630007800e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemAllBodiesAndOblateness        59101695275 58781176800          1                                 +1.00027592630007800e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemAllBodiesAndOblateness        59863012289 59545581700          1                                 +1.00027592630007800e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemAllBodiesAndOblateness_mean   59468984103 59176379333          1                                 +1.00027592630007800e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisSolarSystemAllBodiesAndOblateness_stddev   311380950   312608082          0                                 +1.00027592630007800e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMajorBodiesOnly                   1450373864 1404009000          1                                 154379 steps, +9.99375002672520880e-01 ua, +1.09612948155355540e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMajorBodiesOnly                   1389765715 1357208700          1                                 154379 steps, +9.99375002672520880e-01 ua, +1.09612948155355540e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMajorBodiesOnly                   1367264723 1372808800          1                                 154379 steps, +9.99375002672520880e-01 ua, +1.09612948155355540e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMajorBodiesOnly_mean              1402468101 1378008833          1                                 154379 steps, +9.99375002672520880e-01 ua, +1.09612948155355540e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMajorBodiesOnly_stddev              35097913   19456743          0                                 154379 steps, +9.99375002672520880e-01 ua, +1.09612948155355540e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMinorAndMajorBodies               2176579451 2184014000          1                                 154379 steps, +9.99375002971591320e-01 ua, +1.09612948399733350e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMinorAndMajorBodies               2166758513 2168413900          1                                 154379 steps, +9.99375002971591320e-01 ua, +1.09612948399733350e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMinorAndMajorBodies               2167771077 2168413900          1                                 154379 steps, +9.99375002971591320e-01 ua, +1.09612948399733350e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMinorAndMajorBodies_mean          2170369680 2173613933          1                                 154379 steps, +9.99375002971591320e-01 ua, +1.09612948399733350e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeMinorAndMajorBodies_stddev           4410386    7353958          0                                 154379 steps, +9.99375002971591320e-01 ua, +1.09612948399733350e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeAllBodiesAndOblateness            2190737223 2184014000          1                                 154379 steps, +9.99375002831793260e-01 ua, +1.09612948296115050e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeAllBodiesAndOblateness            2176962828 2168413900          1                                 154379 steps, +9.99375002831793260e-01 ua, +1.09612948296115050e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeAllBodiesAndOblateness            2180941796 2184014000          1                                 154379 steps, +9.99375002831793260e-01 ua, +1.09612948296115050e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeAllBodiesAndOblateness_mean       2182880616 2178813967          1                                 154379 steps, +9.99375002831793260e-01 ua, +1.09612948296115050e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisL4ProbeAllBodiesAndOblateness_stddev        5788077    7353958          0                                 154379 steps, +9.99375002831793260e-01 ua, +1.09612948296115050e+00 ua  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMajorBodiesOnly                  6446595644 6442841300          1                                 750001 steps, +9.99958305820235370e-01 ua, +9.99469624847801300e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMajorBodiesOnly                  6440181708 6442841300          1                                 750001 steps, +9.99958305820235370e-01 ua, +9.99469624847801300e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMajorBodiesOnly                  6433291172 6442841300          1                                 750001 steps, +9.99958305820235370e-01 ua, +9.99469624847801300e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMajorBodiesOnly_mean             6440022841 6442841300          1                                 750001 steps, +9.99958305820235370e-01 ua, +9.99469624847801300e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMajorBodiesOnly_stddev              5432689          0          0                                 750001 steps, +9.99958305820235370e-01 ua, +9.99469624847801300e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMinorAndMajorBodies              10039601302 10046464400          1                                 750001 steps, +9.99958284539249060e-01 ua, +9.99470138411926710e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMinorAndMajorBodies              10045879816 10030864300          1                                 750001 steps, +9.99958284539249060e-01 ua, +9.99470138411926710e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMinorAndMajorBodies              10041628813 10046464400          1                                 750001 steps, +9.99958284539249060e-01 ua, +9.99470138411926710e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMinorAndMajorBodies_mean         10042369977 10041264367          1                                 750001 steps, +9.99958284539249060e-01 ua, +9.99470138411926710e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeMinorAndMajorBodies_stddev           2616222     7353958          0                                 750001 steps, +9.99958284539249060e-01 ua, +9.99470138411926710e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeAllBodiesAndOblateness           10186283326 10186865300          1                                 750001 steps, +9.99958277683878570e-01 ua, +9.99468831450655270e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeAllBodiesAndOblateness           10175492739 10186865300          1                                 750001 steps, +9.99958277683878570e-01 ua, +9.99468831450655270e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeAllBodiesAndOblateness           10179186081 10155665100          1                                 750001 steps, +9.99958277683878570e-01 ua, +9.99468831450655270e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeAllBodiesAndOblateness_mean      10180320715 10176465233          1                                 750001 steps, +9.99958277683878570e-01 ua, +9.99468831450655270e+01 nmi  // NOLINT(whitespace/line_length)
// BM_EphemerisLEOProbeAllBodiesAndOblateness_stddev        4477703    14707915          0                                 750001 steps, +9.99958277683878570e-01 ua, +9.99468831450655270e+01 nmi  // NOLINT(whitespace/line_length)

#include <memory>
#include <vector>

#include "astronomy/frames.hpp"
#include "base/not_null.hpp"
#include "geometry/named_quantities.hpp"
#include "geometry/quaternion.hpp"
#include "geometry/rotation.hpp"
#include "integrators/embedded_explicit_runge_kutta_nyström_integrator.hpp"
#include "integrators/symplectic_runge_kutta_nyström_integrator.hpp"
#include "physics/degrees_of_freedom.hpp"
#include "physics/discrete_trajectory.hpp"
#include "physics/ephemeris.hpp"
#include "physics/massless_body.hpp"
#include "quantities/astronomy.hpp"
#include "quantities/bipm.hpp"
#include "quantities/elementary_functions.hpp"
#include "quantities/numbers.hpp"
#include "quantities/quantities.hpp"
#include "quantities/si.hpp"
#include "testing_utilities/solar_system_factory.hpp"

// Must come last to avoid conflicts when defining the CHECK macros.
#include "benchmark/benchmark.h"

namespace principia {

using astronomy::ICRFJ2000Ecliptic;
using astronomy::ICRFJ2000Equator;
using astronomy::kEquatorialToEcliptic;
using base::not_null;
using geometry::Position;
using geometry::Quaternion;
using geometry::Rotation;
using integrators::DormandElMikkawyPrince1986RKN434FM;
using integrators::McLachlanAtela1992Order5Optimal;
using physics::DegreesOfFreedom;
using physics::DiscreteTrajectory;
using physics::Ephemeris;
using physics::MasslessBody;
using quantities::DebugString;
using quantities::Sqrt;
using quantities::astronomy::JulianYear;
using quantities::bipm::NauticalMile;
using quantities::si::AstronomicalUnit;
using quantities::si::Kilo;
using quantities::si::Metre;
using quantities::si::Milli;
using quantities::si::Minute;
using testing_utilities::SolarSystemFactory;

namespace benchmarks {

namespace {

void EphemerisSolarSystemBenchmark(SolarSystemFactory::Accuracy const accuracy,
                                   not_null<benchmark::State*> const state) {
  Length error;
  while (state->KeepRunning()) {
    state->PauseTiming();
    auto const at_спутник_1_launch =
        SolarSystemFactory::AtСпутник1Launch(accuracy);
    Instant const final_time = at_спутник_1_launch->epoch() + 100 * JulianYear;

    auto const ephemeris =
        at_спутник_1_launch->MakeEphemeris(
            McLachlanAtela1992Order5Optimal<Position<ICRFJ2000Equator>>(),
            45 * Minute,
            5 * Milli(Metre));

    state->ResumeTiming();
    ephemeris->Prolong(final_time);
    state->PauseTiming();
    error = (at_спутник_1_launch->trajectory(
                 *ephemeris, 
                 SolarSystemFactory::name(SolarSystemFactory::kSun)).
                     EvaluatePosition(final_time, nullptr) -
             at_спутник_1_launch->trajectory(
                 *ephemeris,
                 SolarSystemFactory::name(SolarSystemFactory::kEarth)).
                     EvaluatePosition(final_time, nullptr)).
                 Norm();
    state->ResumeTiming();
  }
  state->SetLabel(DebugString(error / AstronomicalUnit) + " ua");
}

void EphemerisL4ProbeBenchmark(SolarSystemFactory::Accuracy const accuracy,
                               not_null<benchmark::State*> const state) {
  Length sun_error;
  Length earth_error;
  int steps;

  auto const at_спутник_1_launch =
      SolarSystemFactory::AtСпутник1Launch(accuracy);
  Instant const final_time = at_спутник_1_launch->epoch() + 100 * JulianYear;

  auto const ephemeris =
      at_спутник_1_launch->MakeEphemeris(
          McLachlanAtela1992Order5Optimal<Position<ICRFJ2000Equator>>(),
          45 * Minute,
          5 * Milli(Metre));

  ephemeris->Prolong(final_time);

  while (state->KeepRunning()) {
    state->PauseTiming();
    // A probe near the L4 point of the Sun-Earth system.
    MasslessBody probe;
    DiscreteTrajectory<ICRFJ2000Equator> trajectory;
    DegreesOfFreedom<ICRFJ2000Equator> const sun_degrees_of_freedom =
        at_спутник_1_launch->initial_state(
            SolarSystemFactory::name(SolarSystemFactory::kSun));
    DegreesOfFreedom<ICRFJ2000Equator> const earth_degrees_of_freedom =
        at_спутник_1_launch->initial_state(
            SolarSystemFactory::name(SolarSystemFactory::kEarth));
    Displacement<ICRFJ2000Ecliptic> const sun_earth_displacement =
        kEquatorialToEcliptic(earth_degrees_of_freedom.position() -
                              sun_degrees_of_freedom.position());
    Rotation<ICRFJ2000Ecliptic, ICRFJ2000Ecliptic> const l4_rotation(
        Quaternion(cos(π / 6), {0, 0, sin(π / 6)}));
    Displacement<ICRFJ2000Ecliptic> const sun_l4_displacement =
        l4_rotation(sun_earth_displacement);
    Velocity<ICRFJ2000Ecliptic> const sun_earth_velocity =
        kEquatorialToEcliptic(earth_degrees_of_freedom.velocity() -
                              sun_degrees_of_freedom.velocity());
    Velocity<ICRFJ2000Ecliptic> const sun_l4_velocity =
        l4_rotation(sun_earth_velocity);
    trajectory.Append(at_спутник_1_launch->epoch(),
                      DegreesOfFreedom<ICRFJ2000Equator>(
                          sun_degrees_of_freedom.position() +
                              kEquatorialToEcliptic.Inverse()(
                                  sun_l4_displacement),
                          sun_degrees_of_freedom.velocity() +
                              kEquatorialToEcliptic.Inverse()(
                                  sun_l4_velocity)));

    state->ResumeTiming();
    ephemeris->FlowWithAdaptiveStep(&trajectory,
                                    1 * Metre,
                                    1 * Metre / Second,
                                    DormandElMikkawyPrince1986RKN434FM<
                                        Position<ICRFJ2000Equator>>(),
                                    final_time);
    state->PauseTiming();

    sun_error = (at_спутник_1_launch->trajectory(
                     *ephemeris, 
                     SolarSystemFactory::name(SolarSystemFactory::kSun)).
                         EvaluatePosition(final_time, nullptr) -
                 trajectory.last().degrees_of_freedom().position()).
                     Norm();
    earth_error = (at_спутник_1_launch->trajectory(
                       *ephemeris, 
                       SolarSystemFactory::name(SolarSystemFactory::kEarth)).
                           EvaluatePosition(final_time, nullptr) -
                   trajectory.last().degrees_of_freedom().position()).
                       Norm();
    steps = trajectory.Times().size();
    state->ResumeTiming();
  }
  std::stringstream ss;
  ss << steps;
  state->SetLabel(ss.str() + " steps, " +
                  DebugString(sun_error / AstronomicalUnit) + " ua, " +
                  DebugString(earth_error / AstronomicalUnit) + " ua");
}

void EphemerisLEOProbeBenchmark(SolarSystemFactory::Accuracy const accuracy,
                                not_null<benchmark::State*> const state) {
  Length sun_error;
  Length earth_error;
  int steps;

  auto const at_спутник_1_launch =
      SolarSystemFactory::AtСпутник1Launch(accuracy);
  Instant const final_time = at_спутник_1_launch->epoch() + 1 * JulianYear;

  auto const ephemeris =
      at_спутник_1_launch->MakeEphemeris(
          McLachlanAtela1992Order5Optimal<Position<ICRFJ2000Equator>>(),
          45 * Minute,
          5 * Milli(Metre));

  ephemeris->Prolong(final_time);

  while (state->KeepRunning()) {
    state->PauseTiming();
    // A probe in low earth orbit.
    MasslessBody probe;
    DiscreteTrajectory<ICRFJ2000Equator> trajectory;
    DegreesOfFreedom<ICRFJ2000Equator> const earth_degrees_of_freedom =
        at_спутник_1_launch->initial_state(
            SolarSystemFactory::name(SolarSystemFactory::kEarth));
    Displacement<ICRFJ2000Equator> const earth_probe_displacement(
        {6371 * Kilo(Metre) + 100 * NauticalMile, 0 * Metre, 0 * Metre});
    Speed const earth_probe_speed =
        Sqrt(at_спутник_1_launch->massive_body(
                 *ephemeris,
                 SolarSystemFactory::name(SolarSystemFactory::kEarth)).
                     gravitational_parameter() /
                     earth_probe_displacement.Norm());
    Velocity<ICRFJ2000Equator> const earth_probe_velocity(
        {0 * Metre / Second, earth_probe_speed, 0 * Metre / Second});
    trajectory.Append(at_спутник_1_launch->epoch(),
                      DegreesOfFreedom<ICRFJ2000Equator>(
                          earth_degrees_of_freedom.position() +
                              earth_probe_displacement,
                          earth_degrees_of_freedom.velocity() +
                              earth_probe_velocity));

    state->ResumeTiming();
    ephemeris->FlowWithAdaptiveStep(&trajectory,
                                    1 * Metre,
                                    1 * Metre / Second,
                                    DormandElMikkawyPrince1986RKN434FM<
                                        Position<ICRFJ2000Equator>>(),
                                    final_time);
    state->PauseTiming();

    sun_error = (at_спутник_1_launch->trajectory(
                     *ephemeris, 
                     SolarSystemFactory::name(SolarSystemFactory::kSun)).
                         EvaluatePosition(final_time, nullptr) -
                 trajectory.last().degrees_of_freedom().position()).
                     Norm();
    earth_error = (at_спутник_1_launch->trajectory(
                       *ephemeris, 
                       SolarSystemFactory::name(SolarSystemFactory::kEarth)).
                           EvaluatePosition(final_time, nullptr) -
                   trajectory.last().degrees_of_freedom().position()).
                       Norm();
    steps = trajectory.Times().size();
    state->ResumeTiming();
  }
  std::stringstream ss;
  ss << steps;
  state->SetLabel(ss.str() + " steps, " +
                  DebugString(sun_error / AstronomicalUnit) + " ua, " +
                  DebugString((earth_error - 6371 * Kilo(Metre)) /
                                  NauticalMile) + " nmi");
}

}  // namespace

void BM_EphemerisSolarSystemMajorBodiesOnly(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisSolarSystemBenchmark(SolarSystemFactory::Accuracy::kMajorBodiesOnly,
                                &state);
}

void BM_EphemerisSolarSystemMinorAndMajorBodies(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisSolarSystemBenchmark(
      SolarSystemFactory::Accuracy::kMinorAndMajorBodies,
      &state);
}

void BM_EphemerisSolarSystemAllBodiesAndOblateness(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisSolarSystemBenchmark(
      SolarSystemFactory::Accuracy::kAllBodiesAndOblateness,
      &state);
}

void BM_EphemerisL4ProbeMajorBodiesOnly(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisL4ProbeBenchmark(SolarSystemFactory::Accuracy::kMajorBodiesOnly,
                            &state);
}

void BM_EphemerisL4ProbeMinorAndMajorBodies(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisL4ProbeBenchmark(SolarSystemFactory::Accuracy::kMinorAndMajorBodies,
                            &state);
}

void BM_EphemerisL4ProbeAllBodiesAndOblateness(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisL4ProbeBenchmark(
      SolarSystemFactory::Accuracy::kAllBodiesAndOblateness,
      &state);
}

void BM_EphemerisLEOProbeMajorBodiesOnly(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisLEOProbeBenchmark(SolarSystemFactory::Accuracy::kMajorBodiesOnly,
                             &state);
}

void BM_EphemerisLEOProbeMinorAndMajorBodies(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisLEOProbeBenchmark(SolarSystemFactory::Accuracy::kMinorAndMajorBodies,
                             &state);
}

void BM_EphemerisLEOProbeAllBodiesAndOblateness(
    benchmark::State& state) {  // NOLINT(runtime/references)
  EphemerisLEOProbeBenchmark(
      SolarSystemFactory::Accuracy::kAllBodiesAndOblateness,
      &state);
}

BENCHMARK(BM_EphemerisSolarSystemMajorBodiesOnly);
BENCHMARK(BM_EphemerisSolarSystemMinorAndMajorBodies);
BENCHMARK(BM_EphemerisSolarSystemAllBodiesAndOblateness);
BENCHMARK(BM_EphemerisL4ProbeMajorBodiesOnly);
BENCHMARK(BM_EphemerisL4ProbeMinorAndMajorBodies);
BENCHMARK(BM_EphemerisL4ProbeAllBodiesAndOblateness);
BENCHMARK(BM_EphemerisLEOProbeMajorBodiesOnly);
BENCHMARK(BM_EphemerisLEOProbeMinorAndMajorBodies);
BENCHMARK(BM_EphemerisLEOProbeAllBodiesAndOblateness);

}  // namespace benchmarks
}  // namespace principia
